### PR TITLE
[FEAT] cleanup GET /wallet endpoint, add wallet api routes

### PR DESCRIPTION
**Pull request by dni**
this removes the functionalitiy to create accounts and wallets via the GET /wallet endpoint in generic.py

it add endpoints inside the api.py for it and the frontend is modified to use the api endpoints

this also simplifies for the `feat/login` for the route.

### CODE REVIEW

File: lnbits/static/js/base.js [hunk 1]
```diff
@@ -64,9 +64,29 @@ window.LNbits = {
         callback
       })
     },
+    createAccount: function (name) {
+      return this.request('post', '/api/v1/wallet', null, {
```
Comment (motorina0) [inline lnbits/static/js/base.js:68]: - the function is called `createAccount` but it creates a wallet

Comment (dni) [reply lnbits/static/js/base.js:68]: you are right! that is abug

File: lnbits/core/views/generic.py [hunk 2]
```diff
@@ -162,77 +161,46 @@ async def extensions_install(
...
-    description="""
-just **wallet_name**: create a new user, then create a new wallet
-                      for user with wallet_name
-just **user_id**: return the first user wallet or create one if none found
-                  (with default wallet_name)
-**user_id** and **wallet_name**: create a new wallet for user with wallet_name
-**user_id** and **wallet_id**: return that wallet if user is the owner
-nothing: create everything
-""",
+    description="show wallet page",
```
Comment (motorina0) [inline lnbits/core/views/generic.py:164]: I hope people are not using this HTML serving endpoint as an REST API one

Comment (motorina0) [reply lnbits/core/views/generic.py:164]: Maybe a similar POST endpoint should exist

File: lnbits/core/static/js/wallet.js [hunk 3]
```diff
@@ -723,27 +723,34 @@ new Vue({
...
       LNbits.api
         .request('PATCH', '/api/v1/wallet', this.g.wallet.adminkey, data)
-        .then(res => {
+        .then(_ => {
           this.$q.notify({
             message: `Wallet updated.`,
             type: 'positive',
             timeout: 3500
           })
-          LNbits.href.updateWallet(
```
Comment (motorina0) [inline lnbits/core/static/js/wallet.js:732]: - this used to work as a redirect `window.location.href`. Is it no longer needed?

File: lnbits/static/js/base.js [hunk 4]
```diff
@@ -64,9 +64,29 @@ window.LNbits = {
...
     },
+    createAccount: function (name) {
+      return this.request('post', '/api/v1/account', null, {
+        name: name
+      })
+    },
     getWallet: function (wallet) {
       return this.request('get', '/api/v1/wallet', wallet.inkey)
     },
+    createWallet: function (name, wallet) {
```
Comment (motorina0) [inline lnbits/static/js/base.js:75]: - param order: `(wallet, name)` feels more natural

File: tests/core/views/test_api.py [hunk 5]
```diff
@@ -25,10 +25,40 @@
...
-    response = await client.get("/")
+async def test_create_account(client):
+    response = await client.post("/api/v1/account", json={"name": "test"})
+    assert response.status_code == 200
+    result = response.json()
+    assert "name" in result
+    assert result["name"] == "test"
+    assert "balance_msat" in result
+    assert "id" in result
+    assert "user" in result
```
Comment (motorina0) [inline tests/core/views/test_api.py:38]: - add test for `patch` (include currency update)

**Comment by motorina0** [comment]
Please `make bundle` and push

**Comment by motorina0** [comment]
- delete wallet does not refresh UI

https://github.com/lnbits/lnbits/assets/2951406/2906e69c-7766-4f2d-a6ea-889d76f59068

**Review by arcbtc — APPROVED** [review]
LFGM. Well spotted!

File: tests/core/views/test_api.py (hunk 5 above)
Comment (dni) [reply tests/core/views/test_api.py:38]: TODO: add test for patch

File: lnbits/core/views/generic.py (hunk 2 above)
Comment (dni) [reply lnbits/core/views/generic.py:164]: we have to annouce it a pbreaking change for `0.11.0`

File: tests/core/views/test_api.py [hunk 6]
```diff
@@ -25,10 +25,40 @@
...
+    assert "balance_msat" in result
+    assert "id" in result
+    assert "adminkey" in result
+    response = await client.delete(
+        "/api/v1/wallet",
+        headers={
+            "X-Api-Key": result["adminkey"],
+            "Content-type": "application/json",
+        },
+    )
```
Comment (callebtc) [inline tests/core/views/test_api.py:61]: Maybe the test can try to create an invoice that is expected to fail from this deleted wallet.

**Review by callebtc — APPROVED** [review]

File: lnbits/core/static/js/wallet.js (hunk 3 above)
Comment (dni) [reply lnbits/core/static/js/wallet.js:736]: it is basically just a reload to change the name, i think its unecessary

File: lnbits/static/js/base.js [hunk 7]
```diff
@@ -64,9 +64,29 @@ window.LNbits = {
...
+      return this.request('post', '/api/v1/wallet', wallet.adminkey, {
+        name: name
+      }).then(res => {
+        window.location = '/wallet?usr=' + res.data.user + '&wal=' + res.data.id
+      })
+    },
+    updateWallet: function (name, wallet) {
```
[the remark this answers is not in the document]
Comment (dni) [reply lnbits/static/js/base.js:82]: there is a test for that.

screenshot-1694524840

**Comment by motorina0** [comment]
**Steps**:
 - delete all wallets of a user

**Expected Result**:
 - a new default wallet is created for the user
 - the user is still valid and can see its extensions and data

**Actual Result**:
 - the user can no longer access the wallet page (or any other page)
 - even if the user has no wallet it might still have some important data in some of the extensions

image

**Comment by dni** [comment]
we could make the `?wal` optional and redirect to the first wallet. do you think its a good idea? and what if there are no wallets left, should we delete the user aswell?

File: lnbits/core/static/js/wallet.js (hunk 3 above)
Comment (dni) [reply lnbits/core/static/js/wallet.js:736]: fixed

File: lnbits/core/crud.py [hunk 8]
```diff
@@ -77,7 +77,7 @@ async def get_user(user_id: str, conn: Optional[Connection] = None) -> Optional[
                 SELECT balance FROM balances WHERE wallet = wallets.id
             ), 0) AS balance_msat
             FROM wallets
-            WHERE "user" = ?
```
Comment (motorina0) [inline lnbits/core/crud.py:80]: - why was this done in this PR, it supposed to be a refactoring only
- this should be done in a different PR

File: lnbits/core/views/generic.py [hunk 9]
```diff
@@ -166,18 +166,25 @@ async def extensions_install(
...
         return template_renderer().TemplateResponse(
             "error.html", {"request": request, "err": "User does not exist."}
         )
 
+    if not wal:
+        if len(user.wallets) == 0:
+            return template_renderer().TemplateResponse(
+                "error.html", {"request": request, "err": "User has no wallets."}
+            )
+        return RedirectResponse(url=f"/wallet?usr={user_id}&wal={user.wallets[0].id}")
```
Comment (motorina0) [inline lnbits/core/views/generic.py:184]: - `user.wallets[0]` can now be `None` as the `user always has a wallet` rule is no longer valid

File: lnbits/core/views/api.py [hunk 10]
```diff
@@ -148,6 +152,30 @@ async def api_update_wallet(
     return await update_wallet(wallet.wallet.id, name, currency)
 
 
+@api_router.delete("/api/v1/wallet")
+async def api_delete_wallet(
+    wallet: WalletTypeInfo = Depends(require_admin_key),
+) -> None:
+    await delete_wallet(
```
Comment (motorina0) [inline lnbits/core/views/api.py:159]: Suggestion:
- if user has zero wallets after the `delete_wallet` operation then create a default empty wallet

File: lnbits/core/views/generic.py (hunk 9 above)
Comment (dni) [reply lnbits/core/views/generic.py:184]: i catch that about with `len(user.wallets) == 0`

File: lnbits/core/views/api.py (hunk 10 above)
Comment (dni) [reply lnbits/core/views/api.py:159]: i do not want to create stuff on `GET` or `DELETE` request

Comment (dni) [reply lnbits/core/views/api.py:159]: i would rather delete the account when the last wallet is deleted

Comment (motorina0) [reply lnbits/core/views/api.py:159]: > i do not want to create stuff on `GET` or `DELETE` request

- I understand for `GET` but for `DELETE` write operations are OK

File: lnbits/core/views/generic.py [hunk 11]
```diff
@@ -162,77 +161,52 @@ async def extensions_install(
...
         )
 
-    logger.debug(
-        "Access "
-        f"{'user '+ user.id + ' ' if user else ''} "
-        f"{'wallet ' + wallet_name if wallet_name else ''}"
-    )
+    if not wal:
+        if len(user.wallets) == 0:
+            wallet = await create_wallet(user_id=user.id)
```
Comment (motorina0) [inline lnbits/core/views/generic.py:181]: - the wallet can **now** be deleted via the API also
- the default wallet should be checked and created at the deletion point
- this page might never be accessed if someone uses LNbits exclusively via the API

File: lnbits/core/crud.py (hunk 8 above)
Comment (dni) [reply lnbits/core/crud.py:80]: else deleted wallets will show up on the wallet page. i directly uses the get_user crud to render it into jinja html template.
